### ISSUE TITLE
Write request hangs in a peculiar case when connection fails

### ISSUE THREAD

**Issue by neoxic** [joyent/libuv#1432]
If a write request is initiated right after a call to `uv_connect` but before `uv_connect_cb` gets fired, it lingers indefinitely until the handle is closed.

Please consider the following test case:

``` c
#include <assert.h>
#include <stdio.h>
#include <uv.h>


uv_loop_t loop;
uv_tcp_t tcp_server;
uv_tcp_t tcp_client;
uv_connect_t crq;
uv_write_t wrq1, wrq2;
uv_buf_t buf = { "HELLO", 4 };


void connection_cb(uv_stream_t *server, int status) {
    printf("connection: %d\n", status);
}

void write_cb(uv_write_t *req, int status) {
    printf("write: %d\n", status);
}

void connect_cb(uv_connect_t *req, int status) {
    printf("connect: %d\n", status);
    //assert(uv_write(&wrq2, (uv_stream_t *)&tcp_client, &buf, 1, write_cb) == 0); /* UNCOMMENT ME */
}

int main() {
    struct sockaddr_in sa;
    assert(uv_ip4_addr("127.0.0.1", 3333, &sa) == 0);
    assert(uv_loop_init(&loop) == 0);
    assert(uv_tcp_init(&loop, &tcp_server) == 0);
    assert(uv_tcp_init(&loop, &tcp_client) == 0);

    //assert(uv_tcp_bind(&tcp_server, (const struct sockaddr *)&sa, 0) == 0); /* UNCOMMENT ME */
    //assert(uv_listen((uv_stream_t *)&tcp_server, 128, connection_cb) == 0); /* UNCOMMENT ME */

    assert(uv_tcp_connect(&crq, &tcp_client, (const struct sockaddr *)&sa, connect_cb) == 0);
    assert(uv_write(&wrq1, (uv_stream_t *)&tcp_client, &buf, 1, write_cb) == 0);

    uv_run(&loop, UV_RUN_DEFAULT);
    return 0;
}
```

The output is:

```
connect: -61
^C
```

The loop neither terminates nor executes the first write request not fails it whereas the connection request has failed.

If the "UNCOMMENT ME" lines get uncommented, the output is:

```
connect: 0
connection: 0
write: 0
write: 0
^C
```

Both writes succeed which is expected and looks fine to me.

**Comment by saghul** [comment]
I think the problem is here: https://github.com/joyent/libuv/blob/master/src/unix/stream.c#L1235 after the connect request callback is called we disarm the fd for writing. We should only do that if the write queue is not empty.

**Comment by txdv** [comment]
I wonder what we should do.

I always call uv_write only once i'm connected, so I never experienced this bug.

But libuv supports queuing of events, if the connection is successful it writes out everything. So if it is not successful, we should flush the write queue with ECANCELED? Or maybe UV_ECONNREFUSED?

Because if I remove the line that @saghul suggest I get UV_EPIPE, which is not a good error message.

**Comment by saghul** [comment]
@txdv I'd go with ECANCELED. We'll need more than just commenting the line, quickly thinking about it: if the connection failed run all queued write callbacks with ECANCELED, else, if there are pending writes don't stop polling for writes.

PS: If any of you is willing to work on this, it will need to be ifdef'd out on Windows, because we currently don't support write queueing there.

**Comment by txdv** [comment]
I'll just add the support in my patch. But I need to get home for windows as usual.

**Comment by saghul** [comment]
Adding support for queing writes on Windows is not part of this issue, please don't add it to the same patch. A different patch which adds support for it is most welcome! :-)

**Comment by txdv** [comment]
@saghul https://github.com/txdv/libuv/compare/tcp-writer-after-connect ?

**Comment by saghul** [comment]
Left some comments. Can you send a PR with it? It's easier to track progress.
